--- click-2.0.1/userlevel/elements.cc	2013-07-26 13:12:48.220000481 -0400
+++ click-2.0.1/userlevel/elements.cc.good	2013-07-26 13:12:43.612000575 -0400
@@ -88,7 +88,9 @@
 #include "../elements/analysis/aggregateip.hh"
 #include "../elements/xia/xiarandomize.hh"
 #include "../elements/xia/iprandomize.hh"
+#ifdef CLICK_ANDROID
+#else
 #include "../elements/userlevel/ipflowrawsockets.hh"
+#endif
 #include "../elements/standard/unqueue.hh"
 #include "../elements/standard/truncate.hh"
@@ -623,7 +625,9 @@
 case 330: return new XIATransport;
 case 331: return new XIAXIDRouteTable;
 case 332: return new XIAXIDTypeClassifier;
 case 333: return new XIAXIDTypeCounter;
 case 334: return new XRoute;
 case 335: return new XTRANSPORT;
+ifdef CLICK_ANDROID
+#else
 case 336: return new IPFlowRawSockets;
+#endif
 default: return 0;
@@ -981,7 +985,9 @@
 (void) click_add_element_type_stable("FromHost", beetlemonkey, 285);
 (void) click_add_element_type_stable("FromRawSocket", beetlemonkey, 286);
 (void) click_add_element_type_stable("FromSocket", beetlemonkey, 287);
+#ifdef CLICK_ANDROID
+#else
 (void) click_add_element_type_stable("IPFlowRawSockets", beetlemonkey, 288);
+#endif
 (void) click_add_element_type_stable("KernelFilter", beetlemonkey, 289);
 (void) click_add_element_type_stable("KernelTap", beetlemonkey, 290);
 (void) click_add_element_type_stable("KernelTun", beetlemonkey, 291);
