diff --git a/click-2.0.1/elements/xia/xtransport.cc b/click-2.0.1/elements/xia/xtransport.cc
index 70ca14c..8b43e42 100644
--- a/click-2.0.1/elements/xia/xtransport.cc
+++ b/click-2.0.1/elements/xia/xtransport.cc
@@ -41,6 +41,7 @@ XTRANSPORT::XTRANSPORT()
 	isConnected = false;
 
 	_ackdelay_ms = ACK_DELAY;
+	_migrateackdelay_ms = ACK_DELAY * 10;
 	_teardown_wait_ms = TEARDOWN_DELAY;
 
 //	pthread_mutexattr_init(&_lock_attr);
@@ -199,7 +200,7 @@ XTRANSPORT::run_timer(Timer *timer)
 				}
 			} else if (daginfo->migrateack_waiting == true && daginfo->expiry <= now ) {
 				//click_chatter("Timer: migrateack waiting\n");
-				if (daginfo->num_migrate_tries <= MAX_CONNECT_TRIES) {
+				if (daginfo->num_migrate_tries <= MAX_MIGRATE_TRIES) {
 
 					//click_chatter("Timer: SYN RETRANSMIT! \n");
 					copy = copy_packet(daginfo->migrate_pkt, daginfo);
@@ -210,7 +211,7 @@ XTRANSPORT::run_timer(Timer *timer)
 
 					daginfo->timer_on = true;
 					daginfo->migrateack_waiting = true;
-					daginfo->expiry = now + Timestamp::make_msec(_ackdelay_ms);
+					daginfo->expiry = now + Timestamp::make_msec(_migrateackdelay_ms);
 					daginfo->num_migrate_tries++;
 				} else {
 					// Stop sending the connection request & Report the failure to the application
@@ -659,6 +660,10 @@ void XTRANSPORT::ProcessNetworkPacket(WritablePacket *p_in)
 				daginfo.hlim = HLIM_DEFAULT;
 				daginfo.next_seqnum = 0;
 				daginfo.ack_num = 0;
+				daginfo.migrateack_waiting = false;
+				daginfo.synack_waiting = false;
+				daginfo.dataack_waiting = false;
+				daginfo.teardown_waiting = false;
 
 				pending_connection_buf.push(daginfo);
 
@@ -2257,7 +2262,7 @@ void XTRANSPORT::Xchangead(unsigned short _sport)
 		// Set timer
 		daginfo->timer_on = true;
 		daginfo->migrateack_waiting = true;
-		daginfo->expiry = Timestamp::now() + Timestamp::make_msec(_ackdelay_ms);
+		daginfo->expiry = Timestamp::now() + Timestamp::make_msec(_migrateackdelay_ms);
 
 		if (! _timer.scheduled() || _timer.expiry() >= daginfo->expiry )
 			_timer.reschedule_at(daginfo->expiry);
