diff --git a/click-2.0.1/elements/scion/tinyxml2.cc b/click-2.0.1/elements/scion/tinyxml2.cc
index 57efc53..62bf007 100644
--- a/click-2.0.1/elements/scion/tinyxml2.cc
+++ b/click-2.0.1/elements/scion/tinyxml2.cc
@@ -1001,38 +1001,38 @@ void XMLAttribute::SetAttribute( const char* v )
 
 void XMLAttribute::SetAttribute( int v )
 {
-	char buf[BUF_SIZE];
-	TIXML_SNPRINTF( buf, BUF_SIZE-1, "%d", v );	
+	char buf[TINYXML2_BUF_SIZE];
+	TIXML_SNPRINTF( buf, TINYXML2_BUF_SIZE-1, "%d", v );	
 	value.SetStr( buf );
 }
 
 
 void XMLAttribute::SetAttribute( unsigned v )
 {
-	char buf[BUF_SIZE];
-	TIXML_SNPRINTF( buf, BUF_SIZE-1, "%u", v );	
+	char buf[TINYXML2_BUF_SIZE];
+	TIXML_SNPRINTF( buf, TINYXML2_BUF_SIZE-1, "%u", v );	
 	value.SetStr( buf );
 }
 
 
 void XMLAttribute::SetAttribute( bool v )
 {
-	char buf[BUF_SIZE];
-	TIXML_SNPRINTF( buf, BUF_SIZE-1, "%d", v ? 1 : 0 );	
+	char buf[TINYXML2_BUF_SIZE];
+	TIXML_SNPRINTF( buf, TINYXML2_BUF_SIZE-1, "%d", v ? 1 : 0 );	
 	value.SetStr( buf );
 }
 
 void XMLAttribute::SetAttribute( double v )
 {
-	char buf[BUF_SIZE];
-	TIXML_SNPRINTF( buf, BUF_SIZE-1, "%f", v );	
+	char buf[TINYXML2_BUF_SIZE];
+	TIXML_SNPRINTF( buf, TINYXML2_BUF_SIZE-1, "%f", v );	
 	value.SetStr( buf );
 }
 
 void XMLAttribute::SetAttribute( float v )
 {
-	char buf[BUF_SIZE];
-	TIXML_SNPRINTF( buf, BUF_SIZE-1, "%f", v );	
+	char buf[TINYXML2_BUF_SIZE];
+	TIXML_SNPRINTF( buf, TINYXML2_BUF_SIZE-1, "%f", v );	
 	value.SetStr( buf );
 }
 
@@ -1631,32 +1631,32 @@ void XMLPrinter::PushAttribute( const char* name, const char* value )
 
 void XMLPrinter::PushAttribute( const char* name, int v )
 {
-	char buf[BUF_SIZE];
-	TIXML_SNPRINTF( buf, BUF_SIZE-1, "%d", v );	
+	char buf[TINYXML2_BUF_SIZE];
+	TIXML_SNPRINTF( buf, TINYXML2_BUF_SIZE-1, "%d", v );	
 	PushAttribute( name, buf );
 }
 
 
 void XMLPrinter::PushAttribute( const char* name, unsigned v )
 {
-	char buf[BUF_SIZE];
-	TIXML_SNPRINTF( buf, BUF_SIZE-1, "%u", v );	
+	char buf[TINYXML2_BUF_SIZE];
+	TIXML_SNPRINTF( buf, TINYXML2_BUF_SIZE-1, "%u", v );	
 	PushAttribute( name, buf );
 }
 
 
 void XMLPrinter::PushAttribute( const char* name, bool v )
 {
-	char buf[BUF_SIZE];
-	TIXML_SNPRINTF( buf, BUF_SIZE-1, "%d", v ? 1 : 0 );	
+	char buf[TINYXML2_BUF_SIZE];
+	TIXML_SNPRINTF( buf, TINYXML2_BUF_SIZE-1, "%d", v ? 1 : 0 );	
 	PushAttribute( name, buf );
 }
 
 
 void XMLPrinter::PushAttribute( const char* name, double v )
 {
-	char buf[BUF_SIZE];
-	TIXML_SNPRINTF( buf, BUF_SIZE-1, "%f", v );	
+	char buf[TINYXML2_BUF_SIZE];
+	TIXML_SNPRINTF( buf, TINYXML2_BUF_SIZE-1, "%f", v );	
 	PushAttribute( name, buf );
 }
 
diff --git a/click-2.0.1/elements/scion/tinyxml2.hh b/click-2.0.1/elements/scion/tinyxml2.hh
index ccc4d6f..8a537f9 100644
--- a/click-2.0.1/elements/scion/tinyxml2.hh
+++ b/click-2.0.1/elements/scion/tinyxml2.hh
@@ -807,7 +807,7 @@ public:
 	void SetAttribute( float value );
 
 private:
-	enum { BUF_SIZE = 200 };
+	enum { TINYXML2_BUF_SIZE = 200 };
 
 	XMLAttribute() : next( 0 ) {}
 	virtual ~XMLAttribute()	{}
@@ -1217,7 +1217,7 @@ private:
 
 	enum {
 		ENTITY_RANGE = 64,
-		BUF_SIZE = 200
+		TINYXML2_BUF_SIZE = 200
 	};
 	bool entityFlag[ENTITY_RANGE];
 	bool restrictedEntityFlag[ENTITY_RANGE];
