diff --git a/click-2.0.1/elements/scion/scioncryptolib.cc b/click-2.0.1/elements/scion/scioncryptolib.cc
index 6b4cc5f..2e1a161 100644
--- a/click-2.0.1/elements/scion/scioncryptolib.cc
+++ b/click-2.0.1/elements/scion/scioncryptolib.cc
@@ -141,7 +141,7 @@ uint32_t SCIONCryptoLib::genMAC(uint8_t* msg, uint16_t msgSize, aes_context* ctx
 #ifdef ENABLE_AESNI 
 	GetCBCOutput( ks, iv, input, buf);
 #else
-	aes_crypt_cbc(ctx, AES_ENCRYPT, msgSize, iv, input, buf);
+	aes_crypt_cbc(ctx, SCION_AES_ENCRYPT, msgSize, iv, input, buf);
 #endif
     uint32_t* ret = (uint32_t*)(buf+msgSize-4);
     return *ret;
diff --git a/click-2.0.1/elements/scion/scioncryptolib.hh b/click-2.0.1/elements/scion/scioncryptolib.hh
index b4bacbe..64a7afe 100644
--- a/click-2.0.1/elements/scion/scioncryptolib.hh
+++ b/click-2.0.1/elements/scion/scioncryptolib.hh
@@ -23,6 +23,11 @@
 #include<polarssl/config.h>
 #include<polarssl/sha1.h>
 #include<polarssl/aes.h>
+/* IPSEC also defines AES_DECRYPT and AES_ENCRYPT */
+const int SCION_AES_DECRYPT = AES_DECRYPT;
+const int SCION_AES_ENCRYPT = AES_ENCRYPT;
+#undef AES_DECRYPT
+#undef AES_ENCRYPT
 #include "define.hh"
 
 // Tenma, add for AESNI
